sdk/couchbase/src/clients/cluster_client.rs

```diff
@@ -0,0 +1,169 @@
+use crate::clients::bucket_client::{
+    BucketClient, BucketClientBackend, Couchbase2BucketClient, CouchbaseBucketClient,
+};
+use crate::clients::query_client::{CouchbaseQueryClient, QueryClient, QueryClientBackend};
+use crate::connstr::{parse, resolve, Address, SrvRecord};
+use crate::error;
+use crate::options::cluster_options::ClusterOptions;
+use couchbase_core::ondemand_agentmanager::{OnDemandAgentManager, OnDemandAgentManagerOptions};
+use std::collections::HashMap;
+use std::sync::Arc;
+use std::time::Duration;
+
+pub(crate) struct ClusterClient {
+    backend: ClusterClientBackend,
+}
+
+enum ClusterClientBackend {
+    CouchbaseClusterBackend(CouchbaseClusterBackend),
+    Couchbase2ClusterBackend(Couchbase2ClusterBackend),
+}
+
+impl ClusterClient {
+    pub async fn connect(
+        conn_str: impl AsRef<str>,
+        opts: ClusterOptions,
+    ) -> error::Result<ClusterClient> {
+        let conn_spec = parse(conn_str)?;
+        let resolved_conn_spec = resolve(conn_spec)?;
+
+        let backend = if let Some(host) = resolved_conn_spec.couchbase2_host {
+            ClusterClientBackend::Couchbase2ClusterBackend(
+                Couchbase2ClusterBackend::connect(host, opts, resolved_conn_spec.options).await?,
+            )
+        } else {
+            ClusterClientBackend::CouchbaseClusterBackend(
+                CouchbaseClusterBackend::connect(
+                    resolved_conn_spec.memd_hosts,
+                    resolved_conn_spec.srv_record,
+                    resolved_conn_spec.use_ssl,
+                    opts,
+                    resolved_conn_spec.options,
+                )
+                .await?,
+            )
+        };
+
+        Ok(ClusterClient { backend })
+    }
+
+    pub async fn bucket_client(&self, name: String) -> error::Result<BucketClient> {
+        match &self.backend {
+            ClusterClientBackend::CouchbaseClusterBackend(ref backend) => {
```

Comment:
For my own understanding, why does this need to be `ref backend`?

Reply:
Because I messed up